File: Tree/Lowest Common Ancestor/1761번 - 정점들의 거리/1761.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define MAX 40001
#define X first
#define Y second

int n, m;
vector<pair<int, int> > edge[MAX];
int parents[MAX], depth[MAX], dist[MAX];
int vis[MAX];


void input() {
    cin >> n;
    for(int i = 0, a, b, c; i < n - 1; i++) {
        cin >> a >> b >> c;
        edge[a].push_back({b, c});
        edge[b].push_back({a, c});
    }
}

void dfs(int node, int dep, int len) {
    vis[node] = true;
    depth[node] = dep;
    dist[node] = len;
    
    for(int i = 0; i < (int)edge[node].size(); i++) {
        int next_node = edge[node][i].X;
        int next_len = len + edge[node][i].Y;
        if(!vis[next_node]) {
            parents[next_node] = node;
            dfs(next_node, dep + 1, next_len);
        }
    }
}

int lca(int node_a, int node_b) {
    int len_a = dist[node_a], len_b = dist[node_b];
    
    if(depth[node_a] > depth[node_b]) swap(node_a, node_b);
    
    while(depth[node_a] != depth[node_b]) {
        node_b = parents[node_b];
    }
    
    while(node_a != node_b) {
        node_a = parents[node_a];
        node_b = parents[node_b];
    }
    
    return len_a + len_b - 2 * dist[node_a];
}

void solve() {
    dfs(1, 0, 0);
    
    cin >> m;
    for(int i = 0, a, b; i < m; i++) {
        cin >> a >> b;
        cout << lca(a, b) << '\n';
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Tree/Lowest Common Ancestor/1761번 - 정점들의 거리/1761.py
import sys
from collections import deque

input = sys.stdin.readline

n = int(input())
edge = list([] for _ in range(n + 1))

# parents, depth, dist
each_val = list([0] * (n + 1) for _ in range(3))
for _ in range(n - 1):
    a, b, c = map(int, input().split())
    edge[a].append([b, c])
    edge[b].append([a, c])

    
def bfs(st):
    vis = [False] * (n + 1)
    
    Q = deque()
    Q.append(st)
    vis[st] = True
    while Q:
        cur = Q.pop()
        
        for nxt, l in edge[cur]:
            if vis[nxt]:
                continue
            each_val[0][nxt] = cur
            each_val[1][nxt] = each_val[1][cur] + 1
            each_val[2][nxt] = each_val[2][cur] + l
            vis[nxt] = True
            Q.append(nxt)
    
def lca(na, nb):
    len_na, len_nb = each_val[2][na], each_val[2][nb]
    
    if each_val[1][na] > each_val[1][nb]:
        na, nb = nb, na
        
    while each_val[1][na] < each_val[1][nb]:
        nb = each_val[0][nb]
        
    while na != nb:
        na = each_val[0][na]
        nb = each_val[0][nb]
    
    return len_na + len_nb - (2 * each_val[2][na])
    
def solve():
    bfs(1)
    
    m = int(input())
    for _ in range(m):
        a, b = map(int, input().split())
        print(lca(a, b))

if __name__ == '__main__':
    solve()